OpenRA.Mods.Cnc/Traits/TransformsNearResources.cs

```diff
@@ -0,0 +1,118 @@
+﻿#region Copyright & License Information
+/*
+ * Copyright 2007-2019 The OpenRA Developers (see AUTHORS)
+ * This file is part of OpenRA, which is free software. It is made
+ * available to you under the terms of the GNU General Public License
+ * as published by the Free Software Foundation, either version 3 of
+ * the License, or (at your option) any later version. For more
+ * information, see COPYING.
+ */
+#endregion
+
+using System;
+using OpenRA.Mods.Common.Activities;
+using OpenRA.Mods.Common.Traits;
+using OpenRA.Traits;
+
+namespace OpenRA.Mods.Cnc.Traits
+{
+	[Desc("Replace with another actor when a resource spawns adjacent.")]
+	public class TransformsNearResourcesInfo : ITraitInfo
+	{
+		[FieldLoader.Require]
+		[ActorReference]
+		public readonly string IntoActor = null;
+
+		public readonly CVec Offset = CVec.Zero;
+
+		[Desc("Don't render the make animation.")]
+		public readonly bool SkipMakeAnims = false;
+
+		[FieldLoader.Require]
+		[Desc("Resource type which triggers the transformation.")]
+		public readonly string Type = null;
+
+		[Desc("Resource density threshold which is required.")]
+		public readonly int Density = 1;
+
+		[Desc("This many adjacent resource tiles are required.")]
+		public readonly int Adjacency = 1;
+
+		[Desc("The median time (in ticks) until the transformation starts.")]
+		public readonly int MedianTicks = 3000;
+
+		public virtual object Create(ActorInitializer init) { return new TransformsNearResources(init.Self, this); }
+	}
+
+	public class TransformsNearResources : ITransformsNearResources, INotifyAddedToWorld, INotifyRemovedFromWorld
+	{
+		readonly TransformsNearResourcesInfo info;
+		readonly ResourceLayer resourceLayer;
+		readonly Actor self;
+		readonly World world;
+		readonly double probability;
+
+		public TransformsNearResources(Actor self, TransformsNearResourcesInfo info)
+		{
+			this.info = info;
+			this.self = self;
+			world = self.World;
+			resourceLayer = world.WorldActor.Trait<ResourceLayer>();
+			probability = (1 - Math.Pow(2, (double)-resourceLayer.Info.NumberOfTicksBetween / info.MedianTicks)) * 10000;
```

Comment:
Are we actually save to use floating point math in simulation code nowadays?

Reply:
Indeed, this will need to be rewritten using integer calculations.